liquid-staking/src/liquidity/remove_liquidity.rs

```diff
@@ -0,0 +1,135 @@
+};
+
+#[multiversx_sc::module]
+pub trait RemoveLiquidityModule:
+    config::ConfigModule
+    + multiversx_sc_modules::default_issue_callbacks::DefaultIssueCallbacksModule
+    + delegation::DelegationModule
+    + liquidity_pool::LiquidityPoolModule
+    + basics::events::EventsModule
+{
+    #[payable("*")]
+    #[endpoint(removeLiquidity)]
+    fn remove_liquidity(&self) {
+        self.blockchain().check_caller_is_user_account();
+        let mut storage_cache = StorageCache::new(self);
+        let caller = self.blockchain().get_caller();
+        let payment = self.call_value().single_esdt();
+
+        require!(
+            self.is_state_active(storage_cache.contract_state),
+            ERROR_NOT_ACTIVE
+        );
+        require!(
+            storage_cache.ls_token_id.is_valid_esdt_identifier(),
+            ERROR_LS_TOKEN_NOT_ISSUED
+        );
+        require!(
+            payment.token_identifier == storage_cache.ls_token_id,
+            ERROR_BAD_PAYMENT_TOKEN
+        );
+        require!(payment.amount > 0, ERROR_BAD_PAYMENT_AMOUNT);
+
+        let egld_to_unstake = self.pool_remove_liquidity(&payment.amount, &mut storage_cache);
+        require!(
+            egld_to_unstake >= MIN_EGLD_TO_DELEGATE,
+            ERROR_INSUFFICIENT_UNSTAKE_AMOUNT
+        );
+        self.burn_ls_token(&payment.amount);
+
+        let delegation_contract = self.get_delegation_contract_for_undelegate(&egld_to_unstake);
+
+        let delegation_contract_mapper = self.delegation_contract_data(&delegation_contract);
+        delegation_contract_mapper
+            .update(|contract_data| contract_data.egld_in_ongoing_undelegation += &egld_to_unstake);
+
+        let gas_for_async_call = self.get_gas_for_async_call();
+        self.tx()
+            .to(delegation_contract.clone())
+            .typed(delegation_proxy::DelegationMockProxy)
+            .undelegate(egld_to_unstake.clone())
+            .gas(gas_for_async_call)
+            .callback(
+                RemoveLiquidityModule::callbacks(self).remove_liquidity_callback(
+                    caller,
+                    delegation_contract,
+                    egld_to_unstake,
+                    payment.amount,
+                ),
+            )
+            .gas_for_callback(MIN_GAS_FOR_CALLBACK)
+            .register_promise();
+    }
+
+    #[promises_callback]
+    fn remove_liquidity_callback(
+        &self,
+        caller: ManagedAddress,
+        delegation_contract: ManagedAddress,
+        egld_to_unstake: BigUint,
+        ls_tokens_to_be_burned: BigUint,
+        #[call_result] result: ManagedAsyncCallResult<()>,
+    ) {
+        let mut storage_cache = StorageCache::new(self);
+        let delegation_contract_mapper = self.delegation_contract_data(&delegation_contract);
+
+        delegation_contract_mapper.update(|contract_data| {
+            contract_data.egld_in_ongoing_undelegation -= &egld_to_unstake;
+        });
```

Comment:
I think it's better to somewhat duplicate this to save some storage operations, otherwise, you have to read and write that big struct twice, more about that in the following suggestion.